tests/integ/datasource/test_oracledb.py

```diff
@@ -0,0 +1,180 @@
+        create_connection_oracledb,
+        table=ORACLEDB_TABLE_NAME,
+        max_workers=4,
+        query_timeout=5,
+    )
+    Utils.check_answer(df, oracledb_real_data)
+
+
+@pytest.mark.parametrize(
+    "create_connection, table_name, expected_result",
+    [
+        (
+            create_connection_oracledb,
+            ORACLEDB_TABLE_NAME,
+            oracledb_real_data,
+        ),
+        (
+            create_connection_oracledb,
+            ORACLEDB_TABLE_NAME_SMALL,
+            oracledb_real_data_small,
+        ),
+    ],
+)
+@pytest.mark.parametrize("fetch_size", [1, 3])
+def test_dbapi_batch_fetch(
+    session, create_connection, table_name, expected_result, fetch_size, caplog
+):
+    with caplog.at_level(logging.DEBUG):
+        df = session.read.dbapi(
+            create_connection, table=table_name, max_workers=4, fetch_size=fetch_size
+        )
+        # we only expect math.ceil(len(expected_result) / fetch_size) parquet files to be generated
+        # for example, 5 rows, fetch size 2, we expect 3 parquet files
+        assert caplog.text.count("Retrieved file from parquet queue") == math.ceil(
+            len(expected_result) / fetch_size
+        )
+        assert df.order_by("ID").collect() == expected_result
+
+
+def test_type_conversion():
+    invalid_type = OracleDBType("ID", "UNKNOWN", None, None, False)
+    with pytest.raises(NotImplementedError, match="oracledb type not supported"):
+        OracledbDriver(create_connection_oracledb, DBMS_TYPE.ORACLE_DB).to_snow_type(
+            [invalid_type]
+        )
+
+
+def test_oracledb_driver_coverage(caplog):
+    oracledb_driver = OracledbDriver(create_connection_oracledb, DBMS_TYPE.ORACLE_DB)
+    conn = oracledb_driver.prepare_connection(oracledb_driver.create_connection(), 0)
+    assert conn.outputtypehandler == output_type_handler
+
+    oracledb_driver.to_snow_type(
+        [OracleDBType("NUMBER_COL", oracledb.DB_TYPE_NUMBER, 40, 2, True)]
+    )
+    assert "Snowpark does not support column" in caplog.text
+
+
+def test_udtf_ingestion_oracledb(session):
+    from tests.parameters import ORACLEDB_CONNECTION_PARAMETERS
+
+    def create_connection_oracledb():
+        import oracledb
+
+        host = ORACLEDB_CONNECTION_PARAMETERS["host"]
+        port = ORACLEDB_CONNECTION_PARAMETERS["port"]
+        service_name = ORACLEDB_CONNECTION_PARAMETERS["service_name"]
+        username = ORACLEDB_CONNECTION_PARAMETERS["username"]
+        password = ORACLEDB_CONNECTION_PARAMETERS["password"]
+        dsn = f"{host}:{port}/{service_name}"
+        connection = oracledb.connect(user=username, password=password, dsn=dsn)
+        return connection
+
+    df = session.read.dbapi(
+        create_connection_oracledb,
+        table="ALL_TYPE_TABLE",
+        udtf_configs={
+            "external_access_integration": ORACLEDB_TEST_EXTERNAL_ACCESS_INTEGRATION
+        },
+    ).order_by("ID")
+    Utils.check_answer(df, oracledb_real_data)
```

Comment:
nit: can we improve this test to verify that when udtf_configs is passed, the code indeed creates a UDTF for ingestion under the hood, rather than falling back to the Parquet file-based approach?